liblangutil/ParserBase.cpp

```diff
@@ -57,30 +57,79 @@ Token ParserBase::advance()
+	else if (TokenTraits::isReservedKeyword(_token))
+		return "reserved keyword '" + TokenTraits::friendlyName(_token) + "'";
+	else if (TokenTraits::isElementaryTypeName(_token)) //for the sake of accuracy in reporting
+	{
+		ElementaryTypeNameToken elemTypeName = m_scanner->currentElementaryTypeNameToken();
+		return "'" + elemTypeName.toString() + "'";
+	}
+	else
+		return "'" + TokenTraits::friendlyName(_token) + "'";
+};
+
 void ParserBase::expectToken(Token _value, bool _advance)
 {
 	Token tok = m_scanner->currentToken();
 	if (tok != _value)
 	{
-		auto tokenName = [this](Token _token)
+		std::string const expectToken = ParserBase::tokenName(_value);
+		if (m_parserErrorRecovery)
+			parserError("Expected " + expectToken + " but got " + tokenName(tok));
+		else
+			fatalParserError("Expected " + expectToken + " but got " + tokenName(tok));
+		// Do not advance so that recovery can sync or make use of the current token. This is especially useful if the expected token
+		// is the only one that is missing and is at the end of a construct.
+		// "{ ... ; }" is such an example.
+		//        ^
+		_advance = false;
+	}
+	if (_advance)
+		m_scanner->next();
+}
+
+void ParserBase::expectTokenOrConsumeUntil(Token _value, char const *_lhs, bool _advance)
+{
+	Token tok = m_scanner->currentToken();
+	if (tok != _value)
+	{
+		Token token = m_scanner->currentToken();
+		int startPosition = position();
+		SourceLocation errorLoc = SourceLocation{startPosition, endPosition(), source()};
+		while (token != _value && token != Token::EOS)
+			token = m_scanner->next();
+		std::string const expectToken = ParserBase::tokenName(_value);
+		std::string const mess = "In <" + string(_lhs) + ">, " + expectToken + "is expected; got " +  ParserBase::tokenName(tok) +  "instead.";
+		if (token == Token::EOS)
 		{
-			if (_token == Token::Identifier)
-				return string("identifier");
-			else if (_token == Token::EOS)
-				return string("end of source");
-			else if (TokenTraits::isReservedKeyword(_token))
-				return string("reserved keyword '") + TokenTraits::friendlyName(_token) + "'";
-			else if (TokenTraits::isElementaryTypeName(_token)) //for the sake of accuracy in reporting
-			{
-				ElementaryTypeNameToken elemTypeName = m_scanner->currentElementaryTypeNameToken();
-				return string("'") + elemTypeName.toString() + "'";
-			}
+			// rollback to where the token started, and raise exception to be caught at a higher level.
+			m_scanner->seek(startPosition);
+			m_inParserRecovery = true;
+			fatalParserError(errorLoc, mess);
+		}
+		else
+		{
+			if (m_inParserRecovery)
+				parserWarning("Recovered in <" + string(_lhs) + "> at " + expectToken + ".");
 			else
-				return string("'") + TokenTraits::friendlyName(_token) + "'";
-		};
-
-		fatalParserError(string("Expected ") + tokenName(_value) + string(" but got ") + tokenName(tok));
+				parserError(errorLoc, mess + "Recovered at next " + expectToken);
+			m_inParserRecovery = false;
```

Comment:
Indentation

Reply:
I don't understand what's wrong here. For things like indentation, I would be grateful if you just do/suggest the commit.

Reply:
Ah sorry, the code is correct. The diff just displayed it a little weirdly.